sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/common/auth/providers/StsAssumeRoleWithDynamicWebIdentityCredentialsProvider.java

```diff
@@ -0,0 +1,233 @@
+
+    abstract String assumedRoleArn();
+
+    abstract String audience();
+
+    abstract String webIdTokenProviderFQCN();
+
+    abstract Integer sessionDurationSecs();
+
+    abstract StsAssumeRoleWithDynamicWebIdentityCredentialsProvider autoBuild(); // not public
+
+    /**
+     * Validates and fully initializes a {@link
+     * StsAssumeRoleWithDynamicWebIdentityCredentialsProvider} instance.
+     *
+     * @return the initialized credentials provider instance.
+     */
+    public StsAssumeRoleWithDynamicWebIdentityCredentialsProvider build() {
+      checkState(audience() != null, "Audience value should not be null");
+      checkState(assumedRoleArn() != null, "The role to assume should not be null");
+      checkState(
+          webIdTokenProviderFQCN() != null,
+          "The web id token provider fully qualified class name should not be null");
+      setCredentialsProviderDelegate(
+          credentialsProviderDelegate()
+              .orElse(
+                  CredentialsProviderDelegate.create(
+                      Suppliers.memoize(() -> WebIdTokenProvider.create(webIdTokenProviderFQCN())),
+                      audience(),
+                      assumedRoleArn(),
+                      sessionDurationSecs())));
+      return autoBuild();
+    }
+  }
+
+  /**
+   * Given the {@link StsAssumeRoleWithWebIdentityCredentialsProvider} is final and can not be
+   * easily mocked for testing purposes, this simple delegate container will be used to simplify
+   * unit testing.
+   */
+  static class CredentialsProviderDelegate {
+
+    private final Supplier<StsAssumeRoleWithWebIdentityCredentialsProvider>
+        credentialsProviderDelegate;
+
+    CredentialsProviderDelegate(
+        Supplier<StsAssumeRoleWithWebIdentityCredentialsProvider> credentialsProviderDelegate) {
+      this.credentialsProviderDelegate = credentialsProviderDelegate;
+    }
+
+    public static CredentialsProviderDelegate create(
+        Supplier<WebIdTokenProvider> webIdTokenProvider,
+        String audience,
+        String assumedRoleArn,
+        @Nullable Integer sessionDurationSecs) {
+      return new CredentialsProviderDelegate(
+          Suppliers.memoize(
+              () ->
+                  createCredentialsDelegate(
+                      webIdTokenProvider, audience, assumedRoleArn, sessionDurationSecs)));
+    }
+
+    public AwsCredentials resolveCredentials() {
+      return credentialsProviderDelegate.get().resolveCredentials();
+    }
+
+    public void close() {
+      credentialsProviderDelegate.get().close();
+    }
+
+    static StsAssumeRoleWithWebIdentityCredentialsProvider createCredentialsDelegate(
+        Supplier<WebIdTokenProvider> webIdTokenProvider,
+        String audience,
+        String assumedRoleArn,
+        @Nullable Integer sessionDurationSecs) {
+      return StsAssumeRoleWithWebIdentityCredentialsProvider.builder()
+          .asyncCredentialUpdateEnabled(true)
+          .refreshRequest(
```

Comment:
does this "refreshRequest" correspond to the functionality that

> An implementation of AwsCredentialsProvider that periodically sends an {@link
> AssumeRoleWithWebIdentityRequest} to the AWS

If so why we really need this wrapper provider instead of setup a StsAssumeRoleWithWebIdentityCredentialsProvider with refreshRequest ?

Reply:
See the [configuration serialization](https://github.com/apache/beam/blob/master/sdks/java/io/amazon-web-services2/src/main/java/org/apache/beam/sdk/io/aws2/options/AwsModule.java#L261), it grabs the generated token and serializes it. When the options are propagated to workers the token value is deserialized and used for authentication, this of course works only for a limited period of time (until the provided token value expires). 

All classes in the STS package are final, so we need to rely on delegates to implement the logic we need here (which is generate a new token for every refresh request from the AWS auth framework).